File: Java/PainterPartition.java
public class Solution {
    public int paint(int A, int B, ArrayList<Integer> C) {
        long total = 0, max = Long.MIN_VALUE;
        for(Integer c : C){
            total += c;
            max = Math.max(max,c);
        }
        long l = max, h = total;
        while(l<h){
            long mid = (l + (h-l)/2);
            long reqPainters = getRequiredPainters(C,mid);
            if(reqPainters <= A) h = mid;
            else l = mid + 1;
        }
        long ans = ((l%10000003)*(B%10000003))%10000003;
        return (int)ans;
    }
    public long getRequiredPainters(ArrayList<Integer> A , long k){
        long total = 0, reqPainters = 1;
        for(Integer a : A){
            total += a;
            if(total > k){
                total = a;
                reqPainters++;
            }
        }
        return reqPainters;
    }
}

File: C++/PainterPartition.cpp
bool ispossible(int A, vector<int> &C, int mid) {
    int n = C.size(), painter = 1;
    int sum = 0;
    for(int i = 0; i < n; i++) {
        sum += C[i];
        if(sum > mid) {
            painter++;
            sum = C[i];
            if(sum > mid || painter > A)
                return false;
        }
    }
    return true;
}

int Solution::paint(int A, int B, vector<int> &C) {
    int high = 0, low = 0, mid;
    int n = C.size();
    for(int i = 0; i < n; i++)
        high = (high + (1LL * C[i]) % 10000003) % 10000003;
    int ans = high;
    while(low <= high) {
        mid = (low + high) / 2;
        if(ispossible(A, C, mid)) {
            ans = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return int((ans * 1LL * B)%10000003);
}
